Initial version

mecha opened the pull request:
Initial version.

Todo:

- [ ] Add `CharPropType` with support for multi-byte characters

----

Provides a base abstract `Struct` class, which is intended to be extended as follows:

```php
class MyStruct extends Struct {
	static protected function propTypes() : array {
		return [/* prop types */];
	}
}
```

----

Provides the `PropType` interface, which is used as a miniature type system.

----

Provides an extensive list of `PropType` implementations that aim to mirror the corresponding types from the PHP type system as much as possible.

In addition, provides some prop types that are not available for class properties with PHP's current type system:

* `callable`
* union
* intersection

----

Provides a `Ty` class with static functions that create prop type instances. This is intended to aid prop type declarations in `Struct::propTypes()`, via a cleaner and chain-able syntax along with instance caching wherever possible to save memory.

mecha committed:
Added test config

mecha committed:
Added property types

mecha committed:
Added struct class

mecha committed:
Added `Ty` helper class

mecha committed:
Added basic usage instructions to readme

mecha committed:
Added enum property type

mecha committed:
Reformated code

mecha committed:
Added enum helper to `Ty`

mecha committed:
Added missing param doc in enum type constructor

mecha committed:
Updated readme

mecha committed:
Added test for object prop type with invalid class

mecha committed:
Added iterable prop type

mecha committed:
Added iterable tests for prop type validity

mecha committed:
Added stringle tests for prop type validity

mecha committed:
Added iterable prop type to readme

mecha committed:
Using named links in readme

mecha committed:
Added test for `Ty::iterable()`

mecha committed:
Added shorthand for array-like prop types

mecha committed:
Can pass any num of parents to `Ty::object`

mecha committed:
`Ty::enum()` is now variadic

mecha committed:
Updated prop types in readme

mecha committed:
Optimized struct data casting

The casting iteration now is now performed on the prop list, rather than
on the input data. This removes the need for `getDefaults()` which
previously iterated over the properties and maps them to their default
values. This change effectively removes that redundant loop over the
properties, making struct construction much more performant.

mecha committed:
Struct ctor and `with()` use different algorithms

This is primarly an optimization for the `with()` method, which does not
need to iterate over all of the struct's prop types. The new algorithm
now only iterates over the given changes. The `castData()` method is now
inlined into the ctor to save an extra function call.

mecha committed:
Added benchmark script and results in readme

mecha committed:
Fixed struct mock

mecha committed:
Added custom prop type

mecha committed:
Added extending example to readme

mecha committed:
Updated readme examples

mecha committed:
Spaced examples in readme

mecha committed:
Changed package short description

mecha committed:
Fixed package name in readme

That's embarrasing!

XedinUnknown commented on src/Struct.php:
```diff
@@ -0,0 +1,223 @@
+    public function with(array $changes)
+    {
+        if (empty($changes)) {
+            return $this;
```
It would be IMHO more intuitive if here you'd return the `$clone` instead.

mecha replied:
Well, there is functionally no difference. Since a struct is immutable, instance does not matter. That's like saying one string `"foo"` is preferred over another instance of string `"foo"`. The string `"foo"` is always `"foo"`; it's an immutable value and has its own identity.

Similarly, here it doesn't matter if the instance is the same one or a clone with no changes. The struct has its own identity, which is its internal data. So to preserve memory and time, I return `$this`.

XedinUnknown commented on src/Struct.php:
```diff
@@ -0,0 +1,223 @@
+
+        $props = $this->getPropTypes();
+        foreach ($changes as $key => $value) {
+            if (!isset($props[$key])) {
```
The kind of stuff that happens here seems to be similar to what's happening in the constructor. I wouldn't recommend creating a new instance of `$this`, instead of `clone`, because that may miss some non-injectable data. But what about adding a `protected` method that does this, and invoke it here and in the constructor?

mecha replied:
[See this commit](https://github.com/Dhii/structs/pull/1/commits/2336726a085ad83ee5007eba270974c4213597e9).

mecha replied:
Re. creating a new instance of `$this`, I tried avoiding being bound to the constructor to let descendants of `Struct` implement their own constructor if they require it.

mecha replied:
Some additional explanation to add onto what I said in the commit message:

The ctor's job is to set the internal data set. That involves iterating the struct's prop types and for each prop type retrieve the corresponding value from the ctor argument, casting it if given and using the default value if not.

The `with()` method however does something different. It iterates over the argument, not the internal prop types, and casts each value using the corresponding prop type.

The algorithms _are_ similar, true, but fundamentally different. It _is_ possible to create a centralized algorithm, but it would make either the ctor or the `with()` method inefficient, depending on what is iterated over. I wasn't willing to make such a compromise.

XedinUnknown commented on src/Struct.php:
```diff
@@ -0,0 +1,223 @@
+        $thisPropsTypes = $this->getPropTypes();
+        $otherPropTypes = $struct->getPropTypes();
+
+        return $this->__data == $struct->__data && $thisPropsTypes == $otherPropTypes;
```
Does loose array comparison actually do this?? :O

mecha replied:
Yeah. I was surprised too. There are tests that cover this, both for successful checks as well as for failures. Works as you'd expect!

XedinUnknown commented on src/Struct.php:
```diff
@@ -0,0 +1,223 @@
+     *
+     * @return array An associative array that maps the property names to their current values.
+     */
+    public function toArray()
```
I'd much rather see this class be `Traversable`.

mecha replied:
Being traversable isn't what structs are supposed to be.

The only reason I added `toArray()` here is because of how frequently the PHP ecosystem uses arrays, and exporting the struct into an array could be useful when working with other packages or PHP extension APIs.

XedinUnknown commented on src/Ty.php:
```diff
@@ -0,0 +1,280 @@
+ *
+ * @since [*next-version*]
+ */
+class Ty
```
I understand, this class has no interface because it's all static, which means you would couple anyhow. But maybe we can have instead a factory with identical instance methods?

mecha replied:
That would require struct declarations to require a factory instance, which is very inconvenient. Plus I'm very much against factories in general.

mecha replied:
Also, in order to be able to do `new MyStruct([...])`, the struct class must be aware of property types statically. That makes it impossible for a factory instance to be used.

XedinUnknown reviewed (changes requested):
Although the idea of interfacing with properties is uncomfortable to me, I can see how having something like this can be useful. The implementation is very good, besides some minor point.

I am convinced that the only proper way to solve this problem would be for PHP to add readonly properties, and allow declaring properties in an interface. But until then, I can really see the use of structs as DTOs which have self-contained validation rules. I can see how something like this can be decorated by a container, increasing its versatility - perhaps something along the lines of what you mention [here][1].

I also realize that while the structs are internally type-safe, consumer code cannot guarantee the type of a retrieved property. Typed properties exist in later versions of PHP, and they would allow that, but I don't know how these approaches can co-exist. I guess this is something to think about later...

[1]: https://github.com/Dhii/containers/pull/7#pullrequestreview-374709472

mecha committed:
Added static `create` method

mecha committed:
Refactored struct methods to be static

mecha committed:
Updated readme and demo files

mecha committed:
Removed static create method

mecha committed:
Added test for prop types cache

mecha committed:
Added missing throws tag for test

mecha committed:
Updated readme

mecha committed:
Removed unused badges from readme

mecha committed:
Derive return type is now a late static type

mecha committed:
Added static method usage in readme and example

mecha committed:
Prop types cache is now a static class field

mecha committed:
New static `create` method that mirrors the ctor

mecha committed:
Updated tests

mecha committed:
Mentioned used memory speed in benchmark test

mecha committed:
Changed tagline

mecha committed:
Array type optionally restricts element types

mecha committed:
Changed static `create` to `fromArray`

mecha committed:
Discouring ctor overriding

mecha committed:
Readme demonstrates struct construction

mecha committed:
Added missing `Ty::mixed()` method

mecha committed:
Some refactoring ...

Struct prop types are no longer cached in a class static var.
This caused extended classes to share the cache, which is bad.

Tests now create anonymous classes to better simulate expected consumer
usage, which is to extend the `Struct` class. As such, the stub was no
longer required.

A mock prop type class was created to facilitate test authoring (since
we can't inject a mock builder into the class - mocked prop types need
to be available before the ctor is called, and anon classes
automatically call the ctor).

Most of the tests have also been revised to be more sensible to real
world expectations.

Finally, the equivalence method is now called `equals()` and only takes
two arguments to reduce complexity. Consumers can easily use this method
with `array_reduce()` to check for equivalence of multiple instances.

mecha committed:
Added `StructPropType`

mecha committed:
Mentioning new struct prop type + example

mecha committed:
Fixed prop types being shared through inheritance
